include/alp/reference/blas2.hpp

```diff
@@ -518,36 +715,350 @@ namespace alp {
 	 * @see alp::operators::internal::Operator for a discussion on when in-place
 	 *      and/or vectorised operations are used.
 	 */
-	template< Descriptor descr = descriptors::no_operation,
+	template<
+		Descriptor descr = descriptors::no_operation,
 		typename InputType, typename InputStructure,
 		typename IOType, typename IOStructure, typename IOView, typename IOImfR, typename IOImfC,
 		class Monoid
 	>
-	RC foldr( const Scalar< InputType, InputStructure, reference > & alpha,
-		Matrix< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > & A,
-		const Monoid & monoid = Monoid(),
-		const typename std::enable_if< ! alp::is_object< InputType >::value && ! alp::is_object< IOType >::value && alp::is_monoid< Monoid >::value, void >::type * const = NULL ) {
+	RC foldr(
+		const Scalar< InputType, InputStructure, reference > &alpha,
+		Matrix< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > &B,
+		const Monoid &monoid = Monoid(),
+		const std::enable_if_t<
+			!alp::is_object< InputType >::value && ! alp::is_object< IOType >::value && alp::is_monoid< Monoid >::value
+		> * const = nullptr
+	) {
 		// static sanity checks
-		//NO_CAST_OP_ASSERT( ( ! ( descr & descriptors::no_casting ) || std::is_same< typename Monoid::D1, IOType >::value ), "alp::foldl",
-		//	"called with a vector x of a type that does not match the first domain "
-		//	"of the given operator" );
-		//NO_CAST_OP_ASSERT( ( ! ( descr & descriptors::no_casting ) || std::is_same< typename Monoid::D2, InputType >::value ), "alp::foldl",
-		//	"called on a vector y of a type that does not match the second domain "
-		//	"of the given operator" );
-		//NO_CAST_OP_ASSERT( ( ! ( descr & descriptors::no_casting ) || std::is_same< typename Monoid::D3, IOType >::value ), "alp::foldl",
-		//	"called on a vector x of a type that does not match the third domain "
-		//	"of the given operator" );
-		(void)alpha;
-		(void)A;
-		(void)monoid;
+		NO_CAST_OP_ASSERT(
+			( ! ( descr & descriptors::no_casting ) || std::is_same< typename Monoid::D1, IOType >::value ),
+			"alp::foldr",
+			"called with a matrix B of a type that does not match the first domain "
+			"of the given operator"
+		);
+		NO_CAST_OP_ASSERT(
+			( ! ( descr & descriptors::no_casting ) || std::is_same< typename Monoid::D2, InputType >::value ),
```

Comment:
`D2 == IOType`

Reply:
This is also wrong in some foldr(s) in sparse case and also in vector versions (blas1.hpp), please correct them as well. @anyzelman could you also confirm that this is wrong in include/graphblas/reference/blas1.hpp ?